File: LeetCode/easy/Min_Stack.cxx
class MinStack {
private:
    std::stack<int> q;
    std::stack<int> minq;
    
public:
    /** initialize your data structure here. */
    MinStack() {

    }
    
    void push(int x) {
        q.push(x);
        if (minq.empty() == false)
        {
            minq.push(std::min(x, minq.top()));
        }
        else
        {
            minq.push(x);
        }
    }
    
    void pop() {
        if (q.empty() == false)
        {
            q.pop();
            minq.pop();
        }
        
    }
    
    int top() {
        if (!q.empty())
            return q.top();
        else
            return -1;
    }
    
    int getMin() {
        if (minq.empty() == false)
            return minq.top();
        else
            return -1;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: LeetCode/medium/Min_Stack.py
class MinStack:

    def __init__(self):
        self.min_stack = []
        self.stack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if len(self.min_stack) == 0 or val <= self.min_stack[-1]:
            self.min_stack.append(val)

    def pop(self) -> None:
        v = self.stack[-1]
        self.stack.pop(-1)
        if self.min_stack[-1] == v:
            self.min_stack.pop(-1)

    def top(self) -> int:
        return self.stack[-1]
        

    def getMin(self) -> int:
        return self.min_stack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
